File: src/frontend/parser/parserContext.ts
// imports 
import { Token, TokenType } from "../lexer/types";

export default class ParserContext {
  public pos: number;
  public tokens: Token[];

  constructor(tokens: Token[], pos: number){
    this.tokens = tokens;
    this.pos = pos;
  }

  peek(): Token {
    return this.tokens[this.pos];
  }

  next(): Token {
    return this.tokens[this.pos + 1];
  }

  eat(): Token {
    return this.tokens[this.pos++];
  }

  endOf(): boolean {
    return this.peek().type === TokenType.EOF;
  }
}  

File: src/frontend/parser/parser_context.cpp
#include <scriptum/frontend/parser/parser_context.hpp>

namespace scriptum {

ParserContext::ParserContext(std::vector<Token> tokens, std::size_t pos)
    : m_tokens(std::move(tokens)), m_pos(pos)
{
}

const Token& ParserContext::peek() const
{
    return m_tokens[m_pos];
}

const Token& ParserContext::next() const
{
    return m_tokens[m_pos + 1];
}

Token ParserContext::eat()
{
    return m_tokens[m_pos++];
}

bool ParserContext::endOf() const
{
    return peek().type == TokenType::Eof;
}

} // namespace scriptum
